Real-time collaboration: Add CollaborationMode SlotFill for editor toolbar

shekharnwagh opened the pull request:
# Add CollaborationMode SlotFill for editor toolbar

## What?

Adds a new `CollaborationMode` SlotFill that lets plugins display collaboration mode indicators (like view/edit state) in the editor toolbar.

## Why?

For real-time collaboration, we need to show users their current mode - whether they can edit or are in view-only mode. This follows the same pattern as `EditorPresence`.

## How?

- Created `CollaborationMode` SlotFill component
- Added slot as first element in the left toolbar section
- Exported as a private API

## Testing Instructions

1. Open the editor and verify it loads without errors
2. Check that toolbar functions normally (no visual changes without a plugin using the slot)

To test with a plugin:

```jsx
const { CollaborationMode } = unlock( editorStore.privateApis );

<CollaborationMode>
    <div>Edit Mode</div>
</CollaborationMode>
```

The indicator should appear as the first item in the left toolbar.

## Screenshots or screencast

N/A - Infrastructure only, no visual changes without a Fill.

shekharnwagh committed:
Add CollaborationMode SlotFill for editor toolbar

Creates a new SlotFill component similar to EditorPresence that allows
plugins to inject collaboration mode indicators (like view/edit state)
into the editor toolbar.

The slot renders as the first element in editor-document-tools__left,
making collaboration status immediately visible to users. This follows
the same pattern as EditorPresence and is exported as a private API
for use by collaboration plugins.

Mamaduka commented on packages/editor/src/components/collaboration-mode/index.tsx:
```diff
@@ -0,0 +1,13 @@
+ */
+import { createSlotFill } from '@wordpress/components';
+
+const { Fill, Slot } = createSlotFill( 'CollaborationMode' );
```
This SlotFill is still public; when the name is a string, anyone who knows it can use it. Please take a look at #67238 for better examples.

Mamaduka commented on packages/editor/src/components/collaboration-mode/index.tsx:
```diff
@@ -0,0 +1,13 @@
+
+const { Fill, Slot } = createSlotFill( 'CollaborationMode' );
+
+/**
+ * Renders the SlotFill for collaboration mode.
+ */
+export const CollaborationMode = Fill;
+
+export { Slot };
```
This exmport can be simplified.

```jsx
const CollaborationModeSlotFill = createSlotFill( 'CollaborationMode' );

export CollaborationModeSlotFill;

// Somewhere else in the app.

<CollaborationModeSlotFill.Slot></CollaborationModeSlotFill.Slot>

<CollaborationModeSlotFill.Fill></CollaborationModeSlotFill.Fill>
```

Mamaduka reviewed (changes requested):
The `editor-document-tools__left` area is getting crowded with recent iterations.

It would be nice to get design feedback here, before we move forward.

P.S. Is there an associated issue for this task?

## Screenshot
<img width="1570" height="116" alt="CleanShot 2025-11-12 at 16 23 05" src="https://github.com/user-attachments/assets/cf97c9de-f14c-47dd-a9fe-3ad6e40a28e9" />

ingeniumed commented:
@Mamaduka - Thanks for the quick review on this PR! This is actually not going into trunk, but into our experimental branch - wpvip/rtc-plugin where we are building out real-time collaboration. Once it goes into that branch, we cherry pick changes and turn them into formal PRs into trunk. It'd be that point that it'd be ready for formal review.

We'll def take your comments into consideration, but not all of them may be solved in this PR. They will be before the formal PR 🙏🏾 

The corresponding issue is https://github.com/WordPress/gutenberg/issues/52593.

shekharnwagh committed:
Use Symbol for CollaborationMode slot name

Changes the CollaborationMode SlotFill to use a Symbol instead of a
string for the slot name. This makes the slot truly private by ensuring
only code with a reference to the exported Fill/Slot components can
access it.

String-based slot names can be accessed by anyone who knows the name.
Using Symbol() prevents unauthorized usage and follows the pattern
established in PR #67238.

shekharnwagh committed:
hide CollaborationModeSlot in distraction-free mode

ingeniumed reviewed (approved):
This is good for our wpvip branch. We can work on it a bit more before our PR into trunk.
